src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+      }
+
+      odb::dbOrientType right_ori = row_ori;
+      if (row_ori == odb::dbOrientType::MX) {
+        right_ori = odb::dbOrientType::R180;
+      } else if (row_ori == odb::dbOrientType::R0) {
+        right_ori = odb::dbOrientType::MY;
+      }
+      buildCell(block, masterr, right_ori, loc_2_x, loc_2_y, endcap_prefix_);
+    }
+  }
+
+  int endcap_count = phy_idx_ - start_phy_idx;
+  logger_->info(utl::TAP, 4, "Inserted {} endcaps.", endcap_count);
+  return endcap_count;
+}
+
+bool Tapcell::isXInRow(const int x, vector<odb::dbRow*>& subrow)
+{
+  for (odb::dbRow* row : subrow) {
+    odb::Rect row_bb;
+    row->getBBox(row_bb);
+    if (x >= row_bb.xMin() && x <= row_bb.xMax()) {
+      return true;
+    }
+  }
+  return false;
 }
 
+int Tapcell::isXCorner(const int x,
+                       vector<odb::dbRow*>& rows_above,
+                       vector<odb::dbRow*>& rows_below)
+{
+  int in_above = isXInRow(x, rows_above);
+  int in_below = isXInRow(x, rows_below);
 
+  if (in_above && !in_below) {
+    return -1;
+  }
+  if (!in_above && in_below) {
+    return 1;
+  }
+  return 0;
 }
+
+odb::dbMaster* Tapcell::pickCornerMaster(int top_bottom,
+                                         odb::dbOrientType ori,
+                                         odb::dbMaster* cnrcap_nwin_master,
+                                         odb::dbMaster* cnrcap_nwout_master,
+                                         odb::dbMaster* endcap_master)
+{
+  if (top_bottom == 1) {
+    if (ori == odb::dbOrientType::MX) {
+      return cnrcap_nwin_master;
+    } else {
+      return cnrcap_nwout_master;
+    }
+  } else if (top_bottom == -1) {
+    if (ori == odb::dbOrientType::R0) {
+      return cnrcap_nwin_master;
+    } else {
+      return cnrcap_nwout_master;
+    }
+  } else {
+    return endcap_master;
+  }
+}
+
+int Tapcell::insertTapcells(vector<vector<odb::dbRow*>>& rows,
+                            string tapcell_master,
+                            int& dist)
+{
+  int start_phy_idx = phy_idx_;
+  odb::dbBlock* block = db_->getChip()->getBlock();
+
+  odb::dbMaster* master = db_->findMaster(tapcell_master.c_str());
+  if (master == nullptr) {
+    logger_->error(utl::TAP, 11, "Master {} not found.", tapcell_master);
+  }
+  int tap_width = master->getWidth();
+  int dist1 = dist * (db_->getTech()->getLefUnits());
+  int dist2 = 2 * dist * (db_->getTech()->getLefUnits());
+
+  int y;
+  std::map<int, vector<vector<int>>> row_fills;
+  for (vector<int>& placement : filled_sites_) {
+    y = placement[0];
+    int x_start = placement[1];
+    int x_end = placement[2];
+    vector<int> x_start_end;
+    x_start_end.push_back(x_start);
+    x_start_end.push_back(x_end);
+    if (row_fills.find(y) == row_fills.end()) {
+      vector<vector<int>> vector2;
+      vector2.push_back(x_start_end);
+      row_fills.insert(std::pair<int, vector<vector<int>>>(y, vector2));
+    } else {
+      row_fills[y].push_back(x_start_end);
+    }
```

Comment:
I think this can just be 

row_fills[y].push_back(x_start_end);

If y isn't present in row_fills it will be inserted by [] and a vector default constructed.